### PR TITLE
[sw/dif] Entropy source implementation

### PR DESCRIPTION
**Pull request by moidx**
Implement the following function calls:

*   `dif_entropy_init()`
*   `dif_entropy_configure()`
*   `dif_entropy_read()`

Add the following test targets:

*   `dif_entropy_unittest`
*   `dif_entropy_smoketest`

Signed-off-by: Miguel Osorio <miguelosorio@google.com>

### CODE REVIEW

**Review by silvestrst — COMMENTED** [review]
@moidx thanks!

This is a quick first pass review.

File: sw/device/lib/dif/dif_entropy.c [hunk 1]
```diff
@@ -4,4 +4,102 @@
...
+  return kDifEntropyOk;
+}
+
+dif_entropy_result_t dif_entropy_configure(const dif_entropy_t *entropy,
+                                           dif_entropy_config_t config) {
+  if (entropy == NULL) {
+    return kDifEntropyBadArg;
+  }
+
+  const uint16_t kMaxLfsrSeed = 15;
```
[the remark this answers is not in the document]
Comment (moidx) [reply sw/device/lib/dif/dif_entropy.c]: Done.

File: sw/device/lib/dif/dif_entropy.c [hunk 2]
```diff
@@ -4,4 +4,102 @@
...
+ */
+static void set_config_register(const dif_entropy_t *entropy,
+                                const dif_entropy_config_t *config) {
+  // TODO: Make this configurable at the API level.
+  uint32_t reg =
+      bitfield_bit32_write(0, ENTROPY_SRC_CONF_BOOT_BYPASS_DISABLE_BIT, 1);
+
+  // Configure test cases
+  reg = bitfield_bit32_write(reg, ENTROPY_SRC_CONF_REPCNT_DISABLE_BIT,
+                             config->tests[kDifEntropyTestRepCount] ? 0 : 1);
```
[the remark this answers is not in the document]
Comment (moidx) [reply sw/device/lib/dif/dif_entropy.c]: Done, updated to `!config->test[]` as the config data encodes inverted logic.

File: sw/device/lib/dif/dif_entropy.c [hunk 3]
```diff
@@ -4,4 +4,102 @@
...
+  }
+
+  uint32_t seed = config.mode == kDifEntropyModeLfsr ? config.lfsr_seed : 0;
+  mmio_region_write32(entropy->params.base_addr, ENTROPY_SRC_SEED_REG_OFFSET,
+                      seed);
+
+  mmio_region_write32(entropy->params.base_addr, ENTROPY_SRC_RATE_REG_OFFSET,
+                      (uint32_t)config.sample_rate);
+
+  // Conditoning bypass is hardcoded to disabled.
```
[the remark this answers is not in the document]
Comment (moidx) [reply sw/device/lib/dif/dif_entropy.c]: Added more details. PTAL.

**Review by moidx — COMMENTED** [review]
Hi @silvestrst, thanks for your comments. PTAL

File: sw/device/tests/dif/dif_entropy_smoketest.c [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+  const dif_entropy_params_t params = {
+    .base_addr = mmio_region_from_addr(TOP_EARLGREY_ENTROPY_SRC_BASE_ADDR),
+  };
+  dif_entropy_t entropy;
+  CHECK(dif_entropy_init(params, &entropy) == kDifEntropyOk);
+
+  const dif_entropy_config_t config = {
+      .mode = kDifEntropyModeLfsr,
+      .tests = {
+          /*kDifEntropyTestRepCount*/ false,
```
Comment (mcy) [inline sw/device/tests/dif/dif_entropy_smoketest.c:32]: In C, you can write `[kDifEntropyTestRepCount] = false,` =D

**Review by mcy — APPROVED** [review]
LG except one nit. =)

**Comment by silvestrst** [comment]
@moidx  I think there is `git clang-format` failure. If you stage your changes and run `git clang-format`, it will fix-it up for you.

LGTM otherwise!

File: sw/device/tests/dif/dif_entropy_smoketest.c (hunk 4 above)
Comment (moidx) [reply sw/device/tests/dif/dif_entropy_smoketest.c:32]: woahh, I love it!

**Comment by moidx** [comment]
> @moidx I think there is `git clang-format` failure. If you stage your changes and run `git clang-format`, it will fix-it up for you.
> 
> LGTM otherwise!

Thanks! Somehow CI is picking an older commit. I need to look into it.

**Pull request merged**